File: location-based-services/jdk-1.8/LocationBasedServices/app/src/main/java/com/example/anhkhue/locationbasedservices/MapsActivity.java
package com.example.anhkhue.locationbasedservices;

import android.Manifest;
import android.content.pm.PackageManager;
import android.content.res.Resources;
import android.location.Address;
import android.location.Geocoder;
import android.location.Location;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.FragmentActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.KeyEvent;
import android.view.inputmethod.EditorInfo;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.Toast;

import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MapStyleOptions;
import com.google.android.gms.maps.model.MarkerOptions;
import com.google.android.gms.tasks.Task;

import java.io.IOException;
import java.util.List;

public class MapsActivity extends FragmentActivity implements OnMapReadyCallback {

    private static final String TAG = MapsActivity.class.getCanonicalName();
    private static final int LOCATION_PERMISSION_REQUEST_CODE = 2204;
    private static final float DEFAULT_ZOOM = 18f;

    private GoogleMap mMap;
    private EditText editSearch;
    private ImageView iconGps;
    private ImageView iconTheme;

    private boolean mLocationPermissionGranted = false;
    private boolean isDark = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        Log.d(TAG, "Create.");

        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_maps);

        SupportMapFragment mapFragment = (SupportMapFragment) getSupportFragmentManager()
                .findFragmentById(R.id.map);
        mapFragment.getMapAsync(this);

        editSearch = findViewById(R.id.editSearch);
        iconGps = findViewById(R.id.iconGps);
        iconTheme = findViewById(R.id.iconTheme);

        askLocationPermission();
    }

    private void askLocationPermission() {
        Log.d(TAG, "Ask location permission.");

        String[] permissions = new String[]{
                Manifest.permission.ACCESS_FINE_LOCATION,
                Manifest.permission.ACCESS_COARSE_LOCATION};

        if (ActivityCompat.checkSelfPermission(this,
                                               Manifest.permission.ACCESS_FINE_LOCATION)
            == PackageManager.PERMISSION_GRANTED ||
            ActivityCompat.checkSelfPermission(this,
                                               Manifest.permission.ACCESS_COARSE_LOCATION)
            == PackageManager.PERMISSION_GRANTED) {

            mLocationPermissionGranted = true;
        } else {
            ActivityCompat.requestPermissions(this,
                                              permissions,
                                              LOCATION_PERMISSION_REQUEST_CODE);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        Log.d(TAG, "Request permission result.");

        mLocationPermissionGranted = false;

        switch (requestCode) {
            case LOCATION_PERMISSION_REQUEST_CODE:
                if (grantResults.length > 0) {
                    for (int grantResult : grantResults) {
                        if (grantResult != PackageManager.PERMISSION_GRANTED) {
                            mLocationPermissionGranted = false;
                            break;
                        }
                    }
                    mLocationPermissionGranted = true;
                    onMapReady(mMap);
                }
                break;
        }
    }

    @Override
    public void onMapReady(GoogleMap googleMap) {
        Log.d(TAG, "Map ready.");

        mMap = googleMap;

        if (mLocationPermissionGranted) {
            getDeviceLocation();
            updateMapUI();
        } else {
            askLocationPermission();
        }
    }

    private void updateMapUI() {
        try {
            mMap.setMyLocationEnabled(true);

            enableMapComponents();

//            mMap.getUiSettings().setCompassEnabled(true);

        } catch (SecurityException e) {
            Log.e(TAG, "SecurityException: " + e.getMessage());
        }
    }

    private void enableMapComponents() {
        editSearch.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId == EditorInfo.IME_ACTION_DONE ||
                actionId == EditorInfo.IME_ACTION_SEARCH ||
                event.getAction() == KeyEvent.KEYCODE_ENTER ||
                event.getAction() == KeyEvent.ACTION_DOWN) {
                searchLocation();
            }
            return false;
        });

        iconGps.setOnClickListener(v -> getDeviceLocation());

        iconTheme.setOnClickListener(v -> {
            if (!isDark) {
                changeTheme(R.raw.dark_theme, R.drawable.ic_gps_dark, R.drawable.ic_light);
            } else {
                changeTheme(R.raw.light_theme, R.drawable.ic_gps, R.drawable.ic_dark);
            }
        });
    }

    private void changeTheme(int theme, int gps, int mode) {
        try {
            boolean success = mMap.setMapStyle(
                    MapStyleOptions.loadRawResourceStyle(this, theme));

            if (success) {
                Log.d(TAG, "Change Theme");

                iconGps.setImageResource(gps);
                iconTheme.setImageResource(mode);

                isDark = true;
            }
        } catch (Resources.NotFoundException e) {
            Log.e(TAG, "Resources.NotFoundException: " + e.getMessage());
        }
    }

    private void searchLocation() {
        Log.d(TAG, "Search location.");

        String searchValue = editSearch.getText().toString();

        Geocoder geocoder = new Geocoder(this);

        try {
            List<Address> addresses = geocoder.getFromLocationName(searchValue, 1);

            if (!addresses.isEmpty()) {
                Address address = addresses.get(0);

                Log.d(TAG, "Location found: " + address.toString());

                moveCamera(new LatLng(address.getLatitude(), address.getLongitude()),
                           DEFAULT_ZOOM,
                           address.getAddressLine(0));
            } else {
                Log.d(TAG, "No location found.");

                Toast.makeText(this,
                               "Unable to find location with keyword '" + searchValue + "'",
                               Toast.LENGTH_SHORT)
                     .show();
            }
        } catch (IOException e) {
            Log.e(TAG, "IOException: " + e.getMessage());
        }
    }

    private void getDeviceLocation() {
        Log.d(TAG, "Get device location.");

        FusedLocationProviderClient mFusedLocationProviderClient =
                LocationServices.getFusedLocationProviderClient(this);

        try {
            Task<Location> locationTask = mFusedLocationProviderClient.getLastLocation();

            locationTask.addOnCompleteListener(task -> {
                if (task.isComplete()) {
                    Log.d(TAG, "Device's location found.");

                    Location location = task.getResult();

                    moveCamera(new LatLng(location.getLatitude(), location.getLongitude()),
                               DEFAULT_ZOOM,
                               "Your location");
                } else {
                    Log.d(TAG, "Device's location not found.");

                    Toast.makeText(getBaseContext(),
                                   "Unable to locate your device.",
                                   Toast.LENGTH_SHORT)
                         .show();
                }
            });
        } catch (SecurityException e) {
            Log.e(TAG, "SecurityException: " + e.getMessage());
        }
    }

    private void moveCamera(LatLng latLng, float zoom, String title) {
        mMap.moveCamera(CameraUpdateFactory.newLatLngZoom(latLng, zoom));

        mMap.addMarker(new MarkerOptions().position(latLng).title(title));
    }
}

File: location-based-services/kotlin/LocationBasedServices/app/src/main/java/com/example/anhkhue/locationbasedservices/MapsActivity.kt
package com.example.anhkhue.locationbasedservices

import android.Manifest
import android.content.pm.PackageManager
import android.content.res.Resources
import android.location.Geocoder
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.KeyEvent
import android.view.inputmethod.EditorInfo
import android.widget.EditText
import android.widget.ImageView
import android.widget.Toast
import com.google.android.gms.location.LocationServices
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MapStyleOptions
import com.google.android.gms.maps.model.MarkerOptions
import java.io.IOException

class MapsActivity : AppCompatActivity(), OnMapReadyCallback {

    companion object {
        const val TAG = "MapsActivity"
        const val LOCATION_PERMISSION_REQUEST_CODE = 2204
        const val DEFAULT_ZOOM = 18f
    }

    private lateinit var mMap: GoogleMap

    private var locationPermissionGranted = false
    private var isDark = false

    private lateinit var editSearch: EditText
    private lateinit var iconGps: ImageView
    private lateinit var iconTheme: ImageView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_maps)

        val mapFragment = supportFragmentManager
                .findFragmentById(R.id.map) as SupportMapFragment
        mapFragment.getMapAsync(this)

        editSearch = findViewById(R.id.editSearch)
        iconGps = findViewById(R.id.ic_gps)
        iconTheme = findViewById(R.id.ic_theme)

        askLocationPermission()
    }

    private fun askLocationPermission() {
        val permissions = arrayOf(Manifest.permission.ACCESS_FINE_LOCATION,
                                  Manifest.permission.ACCESS_COARSE_LOCATION)

        if (ContextCompat.checkSelfPermission(
                        this,
                        Manifest.permission.ACCESS_COARSE_LOCATION)
            == PackageManager.PERMISSION_GRANTED ||
            ContextCompat.checkSelfPermission(
                    this,
                    Manifest.permission.ACCESS_FINE_LOCATION)
            == PackageManager.PERMISSION_GRANTED) {
            locationPermissionGranted = true
        } else {
            ActivityCompat.requestPermissions(this,
                                              permissions,
                                              LOCATION_PERMISSION_REQUEST_CODE)
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int,
                                            permissions: Array<out String>,
                                            grantResults: IntArray) {
        locationPermissionGranted = false
        when (requestCode) {
            LOCATION_PERMISSION_REQUEST_CODE -> {
                if (!grantResults.isEmpty()) {
                    for (grantResult in grantResults) {
                        if (grantResult != PackageManager.PERMISSION_GRANTED) {
                            locationPermissionGranted = false
                            break
                        }
                    }
                    locationPermissionGranted = true
                    onMapReady(mMap)
                } else {
                    askLocationPermission()
                }
            }
        }
    }

    override fun onMapReady(googleMap: GoogleMap) {
        mMap = googleMap

        if (locationPermissionGranted) {
            getDeviceLocation()

            initMapComponents()
        }
    }

    private fun initMapComponents() {
        editSearch.setOnEditorActionListener { _, actionId, event ->
            var handle = false

            if (actionId == EditorInfo.IME_ACTION_DONE ||
                actionId == EditorInfo.IME_ACTION_SEARCH ||
                event.action == KeyEvent.ACTION_DOWN ||
                event.action == KeyEvent.KEYCODE_ENTER) {

                searchLocation()
                handle = true
            }
            handle
        }

        iconGps.setOnClickListener { getDeviceLocation() }

        iconTheme.setOnClickListener {
            if (!isDark) {
                changeTheme(R.raw.dark_theme, R.drawable.ic_light, R.drawable.ic_gps_dark)
                isDark = true
            } else {
                changeTheme()
                isDark = false
            }
        }
    }

    private fun changeTheme(theme: Int = R.raw.light_theme,
                            themeIcon: Int = R.drawable.ic_dark,
                            gpsIcon: Int = R.drawable.ic_gps) {
        try {
            val changed = mMap.setMapStyle(MapStyleOptions.loadRawResourceStyle(this, theme))
            if (changed) {
                Log.i(TAG, "Theme changed.")
                iconGps.setImageResource(gpsIcon)
                iconTheme.setImageResource(themeIcon)
            } else {
                Log.i(TAG, "Unable to parse theme.")
            }
        } catch (e: Resources.NotFoundException) {
            Log.e(TAG, "Resources.NotFoundException: " + e.message)
        }
    }

    private fun searchLocation() {
        val searchValue = editSearch.text.toString()

        val geocoder = Geocoder(this)

        try {
            val addresses = geocoder.getFromLocationName(searchValue, 1)
            if (!addresses.isEmpty()) {
                val address = addresses[0]
                Log.i(TAG, address.toString())

                moveCamera(LatLng(address.latitude, address.longitude),
                           DEFAULT_ZOOM,
                           address.getAddressLine(0))
            } else {
                Log.d(TAG, "Location not found.")

                Toast.makeText(this,
                               "Unable to find location.",
                               Toast.LENGTH_SHORT)
                        .show()
            }
        } catch (e: IOException) {

        }
    }

    private fun getDeviceLocation() {
        val fusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(this)

        try {
            val locationTask = fusedLocationProviderClient.lastLocation

            locationTask.addOnCompleteListener {
                if (it.isSuccessful) {
                    Log.d(TAG, "Location found.")

                    val location = it.result
                    moveCamera(LatLng(location.latitude, location.longitude))
                } else {
                    Log.d(TAG, "Location not found.")
                    Toast.makeText(this,
                                   "Unable to find location",
                                   Toast.LENGTH_SHORT)
                            .show()
                }
            }
        } catch (e: SecurityException) {
            Log.e(TAG, "SecurityException: " + e.message)
        }
    }

    private fun moveCamera(latLng: LatLng,
                           zoom: Float = DEFAULT_ZOOM,
                           title: String = "Your location") {
        mMap.moveCamera(CameraUpdateFactory.newLatLngZoom(latLng, zoom))

        mMap.addMarker(MarkerOptions().position(latLng).title(title))
    }
}
